### PR TITLE
Beta: Case migration

### PR DESCRIPTION
**Pull request by Klaas-Ritense**
Screenshot 2024-03-25 at 14 42 02

### CODE REVIEW

File: document/src/main/kotlin/com/ritense/document/service/DocumentMigrationService.kt [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+                    patch = patch,
+                )
+            }
+
+            val modifiedContent = JsonDocumentContent.build(targetJson, targetJson, targetJsonBuilder.build())
+            Pair(sourceDocument, modifiedContent)
+        }
+    }
+
+    fun applyPatch(
```
Comment (theo-ritense) [inline document/src/main/kotlin/com/ritense/document/service/DocumentMigrationService.kt:150]: There's so much nesting going on in this method that it has become difficult to read. Please refactor and add comments.

File: document/src/test/kotlin/com/ritense/document/service/DocumentMigrationServiceIntTest.kt [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+import com.ritense.document.domain.DocumentMigrationPatch
+import com.ritense.document.domain.DocumentMigrationRequest
+import com.ritense.document.repository.impl.specification.JsonSchemaDocumentSpecificationHelper.Companion.byDocumentDefinitionIdName
+import org.junit.jupiter.api.Test
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.transaction.annotation.Transactional
+import kotlin.test.assertEquals
+
+@Transactional
+class DocumentMigrationServiceIntTest @Autowired constructor(
```
Comment (theo-ritense) [inline document/src/test/kotlin/com/ritense/document/service/DocumentMigrationServiceIntTest.kt:30]: It feels like the logic that is tested here would better fit as unit tests, which are a lot faster.

File: contract/src/main/java/com/ritense/valtimo/contract/json/patch/JsonPatchBuilder.java [hunk 3]
```diff
@@ -144,4 +145,22 @@ public JsonPatch build() {
         return new JsonPatch(operations);
     }
 
+    private JsonNode get(JsonNode json, JsonPointer path) {
```
Comment (theo-ritense) [inline contract/src/main/java/com/ritense/valtimo/contract/json/patch/JsonPatchBuilder.java]: I don't see any changes to `JsonPatchBuilderTest`, so I'm assuming the logic here is not tested.

File: document/src/test/kotlin/com/ritense/document/service/DocumentMigrationServiceIntTest.kt (hunk 2 above)
Comment (Klaas-Ritense) [reply document/src/test/kotlin/com/ritense/document/service/DocumentMigrationServiceIntTest.kt:30]: changed

File: document/src/main/kotlin/com/ritense/document/service/DocumentMigrationService.kt (hunk 1 above)
Comment (Klaas-Ritense) [reply document/src/main/kotlin/com/ritense/document/service/DocumentMigrationService.kt:150]: changed

File: contract/src/main/java/com/ritense/valtimo/contract/json/patch/JsonPatchBuilder.java (hunk 3 above)
Comment (Klaas-Ritense) [reply contract/src/main/java/com/ritense/valtimo/contract/json/patch/JsonPatchBuilder.java]: changed

**Review by theo-ritense — APPROVED** [review]

**Pull request merged**